File: Main/Source/BBC.Dna/SubmitSubbedEntryForm.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Xml;
using BBC.Dna.Component;
using BBC.Dna.Data;
using System.Globalization;

namespace BBC.Dna
{
    /// <summary>
    /// SubmitSubbedEntryForm object
    /// </summary>
    public class SubmitSubbedEntryForm : DnaInputComponent
    {
        bool _isValidEntry = false;
        XmlElement _submitSubbedEntryForm = null;

        /// <summary>
        /// 
        /// </summary>
        /// <param name="context"></param>
        public SubmitSubbedEntryForm(IInputContext context)
            : base(context)
        {
        }

        /// <summary>
        /// Creates the XML for a blank subbed entry form.
        /// </summary>
        public void CreateBlankForm()
        {
            RootElement.RemoveAll();
            _submitSubbedEntryForm = AddElementTag(RootElement, "SUBBED-ENTRY-FORM");
            AddIntElement(_submitSubbedEntryForm, "H2G2ID", 0);
            AddTextTag(_submitSubbedEntryForm, "SUBJECT", "");
            AddTextTag(_submitSubbedEntryForm, "COMMENTS", "");
            XmlElement functions = AddElementTag(_submitSubbedEntryForm, "FUNCTIONS");
            AddTextTag(functions, "FETCH-ENTRY", "");
        }

        /// <summary>
        /// Creates the XML for a recommendation form for the entry with
		///		this entry ID.
        /// </summary>
        public void CreateFromh2g2ID(IUser user, int H2G2ID, string comments)
        {
            bool suitable = true;

            // if daft ih2g2ID value given or user then fail
            if (H2G2ID <= 0 || user.UserID == 0)
            {
                return;
            }
            RootElement.RemoveAll();
            _submitSubbedEntryForm = AddElementTag(RootElement, "SUBBED-ENTRY-FORM");

            if (!GuideEntry.ValidateH2G2ID(H2G2ID))
            {
		        // if ID is invalid then set the flag and put an error message in the XML
                _isValidEntry = false;
                AddErrorXml("INVALID-H2G2ID", H2G2ID.ToString(), _submitSubbedEntryForm);
                return;
            }

            // now check that the h2g2ID is valid and if so get the entry ID from it
            int entryID = H2G2ID / 10;

            using (IDnaDataReader dataReader = InputContext.CreateDnaDataReader("FetchArticleSummary"))
            {
                dataReader.AddParameter("EntryID", entryID);
                dataReader.Execute();

                // Check to see if we found anything
                if (dataReader.HasRows && dataReader.Read())
                {
                    int h2g2ID = dataReader.GetInt32NullAsZero("h2g2ID");
                    string subject = dataReader.GetStringNullAsEmpty("Subject");
                    string editorName = dataReader.GetStringNullAsEmpty("EditorName");
                    int editorID = dataReader.GetInt32NullAsZero("EditorID");
                    int status = dataReader.GetInt32NullAsZero("Status");
                    bool isSubCopy = dataReader.GetBoolean("IsSubCopy");

                    int originalEntryID = dataReader.GetInt32NullAsZero("OriginalEntryID");


                    bool hasSubCopy = dataReader.GetBoolean("HasSubCopy");
                    int subCopyID = dataReader.GetInt32NullAsZero("SubCopyID");

                    int recommendationStatus = dataReader.GetInt32NullAsZero("RecommendationStatus");
                    int recommendedByScoutID = dataReader.GetInt32NullAsZero("RecommendedByScoutID");
                    string scoutName = dataReader.GetStringNullAsEmpty("RecommendedByUserName");

                    string retrievedComments = dataReader.GetStringNullAsEmpty("Comments");

                    // check that this entry is appropriate for recommendation and if not then
                    // assume it is valid then set to invalid if any problems found
                    _isValidEntry = true;
                    // give the reason why
                    // Entry not suitable reasons are:
                    //	- entry is already in edited guide
                    //	- entry is a key article
                    //	- entry is already recommended
                    //	- entry is deleted
                    //	- entry was written by the scout trying to recommend it
                    // only status value 3 (user entry, public) is actually suitable for recommendations

                    string errorType = "WRONG-STATUS";
                    if (!isSubCopy)
                    {
                        _isValidEntry = false;
                        AddErrorXml(errorType, "Not Sub-editors draft copy for Edited Entry", _submitSubbedEntryForm);
                        return;
                    }

                    // if here all okay then put the data into the form
                    AddIntElement(_submitSubbedEntryForm, "H2G2ID", h2g2ID);

                    // only add this data if the entry was suitable for recommending, since it might be
                    // from e.g. a deleted entry and hence should not be accessible
                    if (suitable)
                    {
                        AddTextTag(_submitSubbedEntryForm, "SUBJECT", subject);
                        XmlElement editor = AddElementTag(_submitSubbedEntryForm, "EDITOR");
                        XmlElement userTag = AddElementTag(editor, "USER");
                        AddIntElement(userTag, "USERID", editorID);
                        AddTextTag(userTag, "USERNAME", editorName);
                    }
                    AddTextTag(_submitSubbedEntryForm, "COMMENTS", retrievedComments);
                    // now put in XML saying which functions the form should provide
                    XmlElement functions = AddElementTag(_submitSubbedEntryForm, "FUNCTIONS");
                    AddElementTag(functions, "FETCH-ENTRY");
                    if (_isValidEntry)
                    {
                        AddElementTag(functions, "SUBMIT-ENTRY");
                    }
                }
            }
        }

        /// <summary>
        /// Submits the entry as finished with by the sub editor. Creates appropriate
		///		XML to report whether or not the submission was successful.
        /// </summary>
        public void SubmitSubbedEntry(IUser user, int H2G2ID, string comments)
        {
            bool invalidEntry = false;
            bool userNotSub = false;
            bool okay = false;
            int recommendationStatus = 0;
            DateTime dateReturned;

            // try creating the form for this data
            CreateFromh2g2ID(user, H2G2ID, comments);

            // then check that this was successful and the recommendation was a valid
            // one before proceeding
            if (_isValidEntry)
            {
                // call the SP method that submits an entry that has been subbed and needs to be returned to
				//the editors.
                using (IDnaDataReader dataReader = InputContext.CreateDnaDataReader("SubmitReturnedSubbedEntry"))
                {
                    dataReader.AddParameter("UserID", user.UserID);
                    dataReader.AddParameter("EntryID", H2G2ID / 10);
                    dataReader.AddParameter("Comments", comments);
                    dataReader.Execute();
                    // Check to see if we found anything
                    if (dataReader.HasRows && dataReader.Read())
                    {
                        invalidEntry = dataReader.GetBoolean("InvalidEntry");
                        userNotSub = dataReader.GetBoolean("UserNotSub");
                        okay = dataReader.GetBoolean("Success");
                        recommendationStatus = dataReader.GetInt32NullAsZero("RecommendationStatus");
                        dateReturned = dataReader.GetDateTime("DateReturned");

                        // found invalid even before calling the SP
                        if (!_isValidEntry)
                        {
                            // add some extra error info inside the existing XML
                            AddErrorXml("INVALID-SUBMISSION", "Invalid Submission", _submitSubbedEntryForm);
                        }
                        // entry is not in the accepted recommendations
                        else if (invalidEntry)
                        {
                            // add some extra error info inside the existing XML
                            AddErrorXml("INVALID-ENTRY", "Entry is not in the list of accepted scout recommendations", _submitSubbedEntryForm);
                        }
                        // user is not the sub for this entry
                        else if (userNotSub)
                        {
                            // add an error message saying who recommnded it
                            AddErrorXml("NOT-SUB", "You are not the Sub Editor for this Entry", _submitSubbedEntryForm);
                        }
                        else if (recommendationStatus != 2)
                        {
                            string message;
                            // get the date in a simple format
                            CultureInfo culture = new CultureInfo("en-GB");      
                            switch (recommendationStatus)
                            {
                                case 0: message = "Recommendation has no status";
                                    break;
                                case 1: message = "Recommendation has not yet been allocated to a Sub Editor";
                                    break;
                                case 3: message = "Recommendation has already been submitted on " + dateReturned.ToString("dd-MM-yyyy", culture);
                                    break;
                                default: message = "Recommendation has wrong status";
                                    break;
                            }
                            AddErrorXml("WRONG-STATUS", message, _submitSubbedEntryForm);
                        }
                        // something unspecified went wrong
                        else if (!okay)
                        {
                            AddErrorXml("UNKNOWN", "An unspecified error occurred whilst processing the request.", _submitSubbedEntryForm);
                        }
                        // otherwise all okay!
                        else
                        {
                            // if submission successful then want to return a blank form so
                            // that they can submit more if they wish, so destroy current contents
                            // and create some more
                            CreateBlankForm();

                            // also want to put some XML in to indicate that recommendation was successful though
                            // stylesheet can then put whatever message is appropriate in
                            XmlElement submitted = AddElementTag(_submitSubbedEntryForm, "SUBMITTED");
                            AddAttribute(submitted, "H2G2ID", H2G2ID);
                        }                 
                    }
                }
            }
        }
    }
}

File: Main/Source/Ripley/SubmitSubbedEntryForm.cpp
// SubmitSubbedEntryForm.cpp: implementation of the CSubmitSubbedEntryForm class.
//
//////////////////////////////////////////////////////////////////////

/*

Copyright British Broadcasting Corporation 2001.

This code is owned by the BBC and must not be copied, 
reproduced, reconfigured, reverse engineered, modified, 
amended, or used in any other way, including without 
limitation, duplication of the live service, without 
express prior written permission of the BBC.

The code is provided "as is" and without any express or 
implied warranties of any kind including without limitation 
as to its fitness for a particular purpose, non-infringement, 
compatibility, security or accuracy. Furthermore the BBC does 
not warrant that the code is error-free or bug-free.

All information, data etc relating to the code is confidential 
information to the BBC and as such must not be disclosed to 
any other party whatsoever.

*/


#include "stdafx.h"
#include "SubmitSubbedEntryForm.h"
#include "InputContext.h"
#include "GuideEntry.h"
#include "TDVAssert.h"
#include "StoredProcedure.h"

//////////////////////////////////////////////////////////////////////
// Construction/Destruction
//////////////////////////////////////////////////////////////////////

CSubmitSubbedEntryForm::CSubmitSubbedEntryForm(CInputContext& inputContext) :
	CXMLObject(inputContext),
	m_bIsValidEntry(false)
{
	// no further construction required
}

CSubmitSubbedEntryForm::~CSubmitSubbedEntryForm()
{
	// no further destruction required
}

/*********************************************************************************

	bool CSubmitSubbedEntryForm::CreateBlankForm(CUser* pUser)

	Author:		Kim Harries
	Created:	15/12/2000
	Inputs:		-
	Outputs:	-
	Returns:	true for success or false for failure
	Purpose:	Creates the XML for a blank form.

*********************************************************************************/

bool CSubmitSubbedEntryForm::CreateBlankForm(CUser* pUser)
{
	TDVASSERT(m_pTree == NULL, "CSubmitSubbedEntryForm::CreateBlankForm() called with non-NULL m_pTree");
	TDVASSERT(pUser != NULL, "CSubmitSubbedEntryForm::CreateBlankForm() called with NULL pUser");

	// fail is no user object provided
	if (pUser == NULL)
	{
		return false;
	}
	// if object is not empty then simply delete the existing tree rather than failing
	if (m_pTree != NULL)
	{
		delete m_pTree;
		m_pTree = NULL;
	}

	// variables we will need
	CTDVString	sFormXML = "<SUBBED-ENTRY-FORM>";

	sFormXML << "<H2G2ID>0</H2G2ID>";
	sFormXML << "<SUBJECT></SUBJECT>";
	sFormXML << "<COMMENTS></COMMENTS>";
	sFormXML << "<FUNCTIONS><FETCH-ENTRY/></FUNCTIONS>";
	sFormXML << "</SUBBED-ENTRY-FORM>";

	// create the XML tree from this text
	return CXMLObject::CreateFromXMLText(sFormXML);
}

/*********************************************************************************

	bool CSubmitSubbedEntryForm::CreateFromh2g2ID(CUser* pUser, int iEntryID)

	Author:		Kim Harries
	Created:	15/12/2000
	Inputs:		-
	Outputs:	-
	Returns:	true for success or false for failure
	Purpose:	Creates the XML for a subbed entry submission form for the entry with
				this entry ID.

*********************************************************************************/

bool CSubmitSubbedEntryForm::CreateFromh2g2ID(CUser* pUser, int ih2g2ID)
{
	TDVASSERT(m_pTree == NULL, "CSubmitSubbedEntryForm::CreateBlankForm(...) called with non-NULL m_pTree");
	TDVASSERT(pUser != NULL, "CSubmitSubbedEntryForm::CreateBlankForm(...) called with NULL pUser");
	TDVASSERT(ih2g2ID > 0, "CSubmitSubbedEntryForm::CreateBlankForm(...) called with non-positive ih2g2ID");

	// if daft ih2g2ID value given or a NULL user object then fail
	if (ih2g2ID <= 0 || pUser == NULL)
	{
		return false;
	}
	// if object is not empty then simply delete the existing tree rather than failing
	if (m_pTree != NULL)
	{
		delete m_pTree;
		m_pTree = NULL;
	}

	// variables we will need
	CStoredProcedure	SP;
	CTDVString			sFormXML;
	CTDVString			sComments;
	CTDVString			sSubject;
	CTDVString			sEditorName;
	int					iUserID = pUser->GetUserID();
	int					iEntryID = 0;
	int					iEditorID = 0;
	int					iStatus = 0;
	bool				bIsSubCopy = false;
	bool				bHasSubCopy = false;
	bool				bSuccess = true;
	bool				bInvalidID = false;

	bSuccess = bSuccess && m_InputContext.InitialiseStoredProcedureObject(&SP);
	// start the form element
	sFormXML = "<SUBBED-ENTRY-FORM>";
	// now check that the h2g2ID is valid and if so get the entry ID from it
	if (CGuideEntry::IsValidChecksum(ih2g2ID))
	{
		iEntryID = ih2g2ID / 10;
	}
	else
	{
		// if ID is invalid then set the flag and put an error message in the XML
		// set bSuccess false for time being so that other stuff is not attempted
		TDVASSERT(false, "Invalid h2g2ID in CSubmitSubbedEntryForm::CreateFromh2g2ID(...)");
		bInvalidID = true;
		bSuccess = false;
		m_bIsValidEntry = false;
		sFormXML << "<ERROR TYPE='INVALID-H2G2ID'>" << ih2g2ID << "</ERROR>";
	}
	// get the entry details to go in the form
	bSuccess = bSuccess && SP.FetchArticleSummary(iEntryID, &ih2g2ID, &sSubject, &sEditorName, &iEditorID, &iStatus, &bIsSubCopy, NULL, &bHasSubCopy, NULL, NULL, NULL, NULL, &sComments);
	if (bSuccess)
	{
		// check that this entry is a subs copy of a recommended entry
		m_bIsValidEntry = true;
		if (!bIsSubCopy)
		{
			m_bIsValidEntry = false;
			sFormXML << "<ERROR TYPE='WRONG-STATUS'>Not Sub-editors draft copy for Edited Entry</ERROR>";
		}
	}

	// if all okay then put the data into the form
	if (bSuccess)
	{
		sFormXML << "<H2G2ID>" << ih2g2ID << "</H2G2ID>";
		EscapeAllXML(&sSubject);
		sFormXML << "<SUBJECT>" << sSubject << "</SUBJECT>";
		sFormXML << "<EDITOR>";
		sFormXML << "<USER><USERID>" << iEditorID << "</USERID>";
		sFormXML << "<USERNAME>" << sEditorName << "</USERNAME></USER>";
		sFormXML << "</EDITOR>";
		EscapeAllXML(&sComments);
		sFormXML << "<COMMENTS>" << sComments << "</COMMENTS>";
	}
	// now put in XML saying which functions the form should provide
	sFormXML << "<FUNCTIONS>";
	sFormXML << "<FETCH-ENTRY/>";
	if (m_bIsValidEntry)
	{
		sFormXML << "<SUBMIT-ENTRY/>";
	}
	sFormXML << "</FUNCTIONS>";
	sFormXML << "</SUBBED-ENTRY-FORM>";
	// if only reason we failed was because of invalid ID then try to succeed
	// with just the error message included, that way the stylesheet can decide
	// how best to process it
	if (bInvalidID)
	{
		bSuccess = true;
	}
	// now parse this XML into our tree
	bSuccess = bSuccess && CXMLObject::CreateFromXMLText(sFormXML);
	// log an error if object not created okay
	TDVASSERT(bSuccess, "CSubmitSubbedEntryForm::CreateFromEntryID(...) failed");
	// return the success value
	return bSuccess;
}

/*********************************************************************************

	bool CSubmitSubbedEntryForm::SubmitSubbedEntry(CUser* pViewer, int ih2g2ID, const TDVCHAR* pcComments)

	Author:		Kim Harries
	Created:	15/12/2000
	Inputs:		pViewer - the user (sub editor) returning this entry
				ih2g2ID - the ID of the entry they want to submit as subbed
				pcComments - any comments to be passed on to the editors.
	Outputs:	-
	Returns:	true for success or false for failure
	Purpose:	Submits the entry as finsihed with by the sub editor. Creates appropriate
				XML to report whether or not the submission was successful.

*********************************************************************************/

bool CSubmitSubbedEntryForm::SubmitSubbedEntry(CUser* pViewer, int ih2g2ID, const TDVCHAR* pcComments)
{
	TDVASSERT(pViewer != NULL, "NULL pViewer in CSubmitSubbedEntryForm::SubmitSubbedEntry(...)");

	bool				bInvalidEntry = false;
	bool				bUserNotSub = false;
	bool				bOkay = false;
	int					iRecommendationStatus = 0;
	CTDVDateTime		dtDateReturned;
	bool				bSuccess = true;

	// try creating the form for this data
	bSuccess = bSuccess && CreateFromh2g2ID(pViewer, ih2g2ID);
	// then check that this was successful and the submission is a valid
	// one before proceeding
	if (bSuccess && m_bIsValidEntry)
	{
		CStoredProcedure	SP;
		

		if (!m_InputContext.InitialiseStoredProcedureObject(&SP))
		{
			// if couldn't create SP then log error and fail
			TDVASSERT(false, "Failed to create SP in CSubmitSubbedEntryForm::SubmitSubbedEntry(...)");
			bSuccess = false;
		}
		else
		{
			// call the SP method to send the scouts submission, then fetch any fields returned
			bSuccess = bSuccess && SP.SubmitSubbedEntry(pViewer->GetUserID(), ih2g2ID / 10, pcComments);
			if (bSuccess)
			{
				bInvalidEntry = SP.GetBoolField("InvalidEntry");
				bUserNotSub = SP.GetBoolField("UserNotSub");
				bOkay = SP.GetBoolField("Success");
				iRecommendationStatus = SP.GetIntField("RecommendationStatus");
				dtDateReturned = SP.GetDateField("DateReturned");
			}
		}

	}
	// check how things went and give a report
	if (bSuccess)
	{
		// found invalid even before calling the SP
		if (!m_bIsValidEntry)
		{
			// add some extra error info inside the existing XML
			bSuccess = bSuccess && AddInside("SUBBED-ENTRY-FORM", "<ERROR TYPE='INVALID-SUBMISSION'>Invalid Submission</ERROR>");
		}
		// entry is not in the accepted recommendations
		else if (bInvalidEntry)
		{
			// add some extra error info inside the existing XML
			bSuccess = bSuccess && AddInside("SUBBED-ENTRY-FORM", "<ERROR TYPE='INVALID-ENTRY'>Entry is not in the list of accepted scout recommendations</ERROR>");
		}
		// user is not the sub for this entry
		else if (bUserNotSub)
		{
			bSuccess = bSuccess && AddInside("SUBBED-ENTRY-FORM", "<ERROR TYPE='NOT-SUB'>You are not the Sub Editor for this Entry</ERROR>");
		}
		else if (iRecommendationStatus != 2)
		{
			CTDVString	sMessage;
			switch (iRecommendationStatus)
			{
				case 0: sMessage = "<ERROR TYPE='WRONG-STATUS'>Recommendation has no status</ERROR>";
						break;
				case 1: sMessage = "<ERROR TYPE='WRONG-STATUS'>Recommendation has not yet been allocated to a Sub Editor</ERROR>";
						break;
				case 3: sMessage = "<ERROR TYPE='WRONG-STATUS'>Recommendation has already been submitted on ";
						sMessage << (LPCSTR)dtDateReturned.Format("%d-%m-%Y") << "</ERROR>";
						break;
				default: sMessage = "<ERROR TYPE='WRONG-STATUS'>Recommendation has wrong status</ERROR>";
						break;
			}
			bSuccess = bSuccess && AddInside("SUBBED-ENTRY-FORM", sMessage);
		}
		// something unspecified went wrong
		else if (!bOkay)
		{
			bSuccess = bSuccess && AddInside("SUBBED-ENTRY-FORM", "<ERROR TYPE='UNKNOWN'>An unspecified error occurred whilst processing the request.</ERROR>");
		}
		// otherwise all okay!
		else
		{
			// if submission successful then want to return a blank form so
			// that they can submit more if they wish, so destroy current contents
			// and create some more
			Destroy();
			bSuccess = bSuccess && CreateBlankForm(pViewer);
			// also want to put some XML in to indicate that recommendation was successful though
			// stylesheet can then put whatever message is appropriate in
			CTDVString	sMessage = "<SUBMITTED H2G2ID='";
			sMessage << ih2g2ID << "'/>";
			bSuccess = bSuccess && AddInside("SUBBED-ENTRY-FORM", sMessage);
		}
	}
	// return success value
	return bSuccess;
}
